### PR TITLE
[MonologBridge] Add support for Monolog 3

### PR DESCRIPTION
**Pull request by Seldaek**
| Q             | A
| ------------- | ---
| Branch?       | 6.1
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | 
| License       | MIT
| Doc PR        | symfony/symfony-docs#... 

Adds support for the upcoming version of Monolog. I mostly wanted to do this before releasing it to ensure it is possible to integrate with it and Monolog 2 using the same codebase with only minimal amounts of version-based forks. 

I think the result is pretty good, so I will try and keep pushing for a beta release soon. I still need to add a SymfonyMailerHandler to replace the SwiftMailerHandler which got axed though.

To try it with Monolog 3 one needs to switch the requirement to `^3@dev` for now, but I did that locally and all tests pass with v3.

### CODE REVIEW

File: src/Symfony/Bridge/Monolog/Formatter/ConsoleFormatter.php
Comment (derrabus) [inline src/Symfony/Bridge/Monolog/Formatter/ConsoleFormatter.php]: This change is technically a BC break because the method is not final and thus could be overridden downstream.

**Comment by stof** [comment]
The change being a BC break for classes extending Symfony handlers is also the reason why Symfony 4.x never added support for Monolog 2.
I suggest that in Symfony 6.x (probably 6.2 due to being after the 6.1 feature freeze), we tag all our handlers and formatters as `@final` (or at least the methods of those handlers implementing the Monolog API if they are meant to provide their own inheritance-based extension point) to avoid that kind of blockages again in the future.
And we will need to wait for Symfony 7.0 to support Monolog 3.

**Comment by derrabus** [comment]
> And we will need to wait for Symfony 7.0 to support Monolog 3.

That would be a very unfortunate timeframe. We could work with conditional traits to achieve compatibility for Monolog 2 + 3 and maintain BC if Monolog 2 is installed. Ideally, we could even drop Monolog 2 support in Symfony 7.0 then.

And yes, implementations of external interfaces should be final.

**Comment by Seldaek** [comment]
> That would be a very unfortunate timeframe. We could work with conditional traits to achieve compatibility for Monolog 2 + 3 and maintain BC if Monolog 2 is installed. Ideally, we could even drop Monolog 2 support in Symfony 7.0 then.

Indeed waiting for Symfony 7 would kinda suck, especially as Symfony 6.1 bumping PHP to 8.1 is what motivated me to do the same in Monolog and take advantage of readonly props & enums..

Dropping Monolog 1/2 in Symfony 7 though absolutely would be a good idea.

Anyway in https://github.com/symfony/symfony/pull/46153/commits/22b7ae61806565058b5511aeda061e982dedcc29 I added compatibility traits to get around BC break.

File: src/Symfony/Bridge/Monolog/Processor/DebugProcessor.php
Comment (Seldaek) [inline src/Symfony/Bridge/Monolog/Processor/DebugProcessor.php]: I marked this internal instead of marking the class final as it is being extended in the tests, so maybe it was meant to be extended? I am not sure what the intent was.

File: src/Symfony/Bridge/Monolog/Handler/ConsoleHandler.php
Comment (derrabus) [inline src/Symfony/Bridge/Monolog/Handler/ConsoleHandler.php]:
```suggestion
 * @final since Symfony 6.1
```

If we softly flag a class as `@final` or `@internal` in a minor version, we usually add the Symfony version that introduced that change. This reminds us that we cannot treat the method as final/internal until the next major release.

**Comment by Seldaek** [comment]
OK tests pass now on Monolog 1/2/3 for me. The psalm errors are bogus and should partially be resolved once it can run against Monolog 3.

File: src/Symfony/Bridge/Monolog/CHANGELOG.md
Comment (nicolas-grekas) [inline src/Symfony/Bridge/Monolog/CHANGELOG.md]: s/Added/Add + no final dot

File: src/Symfony/Bridge/Monolog/Processor/AbstractTokenProcessor.php
Comment (nicolas-grekas) [inline src/Symfony/Bridge/Monolog/Processor/AbstractTokenProcessor.php:24]: shouldn't be internal as this abstract class is meant to be used also by userland to me

File: src/Symfony/Bridge/Monolog/Processor/CompatibilityProcessor.php
Comment (nicolas-grekas) [inline src/Symfony/Bridge/Monolog/Processor/CompatibilityProcessor.php:23]: Shouldn't we make all the traits NON-internal, to make it easy for third parties to provide monolog 1+2+3 compat?

**Review by nicolas-grekas — COMMENTED** [review]
What's the plan for Symfony 7? Provide support for only monolog 3? We should then figure out a way to throw deprecations when monolog 1/2 is used (once monolog 3 is out)

File: src/Symfony/Bridge/Monolog/Processor/AbstractTokenProcessor.php
Comment (derrabus) [reply src/Symfony/Bridge/Monolog/Processor/AbstractTokenProcessor.php:24]: Can we finalize the `__invoke()` method instead? It is the implementation of an external contract and not being able to change its signature was one of the challanges we've faced here.

Comment (nicolas-grekas) [reply src/Symfony/Bridge/Monolog/Processor/AbstractTokenProcessor.php:24]: if CompatibilityProcessor is not internal as I suggested, then ppl should use it and implement doInvoke instead,
this would solve the issue without requiring final/internal

File: src/Symfony/Bridge/Monolog/Processor/CompatibilityProcessor.php
Comment (Seldaek) [reply src/Symfony/Bridge/Monolog/Processor/CompatibilityProcessor.php:23]: IMO most people can easily add the new LogRecord or simply do __invoke($record) without type hints and have it work with everything. This is only a problem in symfony due to the strict BC policy and I suspect most impl can get away with just changing this or supporting only monolog 3 going forward.. So I don't think it's worth having this stuff in the public api, but if the core team consensus is to do it I can update the PR.

**Comment by Seldaek** [comment]
Yes I think I'd add a deprecation for monolog 1/2 in 6.2 or 6.3 latest. Doing it in 6.1 would be a bit early probably. I'll definitely try to get a release of monolog out this week anyway, I hope this PR can be finalized/merged soon as well. 

It should be fairly easy to do in the if/else we have for the traits, no?

**Review by nicolas-grekas — APPROVED** [review]
Let's ignore my comment about `@internal`: while making things internal might be too closed (think open/closed principle), we can wait for actual feedback from userland before considering it is ;)

:+1:

**Review by derrabus — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
Thank you @Seldaek.

**Pull request merged**
